contracts/contracts/strategies/balancer/BalancerMetaPoolStrategy.sol

```diff
@@ -0,0 +1,477 @@
+                if (poolAssets[i] == mappedAssets[j]) {
+                    amountsIn[i] = mappedAmounts[j];
+                }
+            }
+        }
+
+        uint256 minBPT = getBPTExpected(_assets, _amounts);
+        uint256 minBPTwSlippage = minBPT.mulTruncate(1e18 - maxDepositSlippage);
+
+        /* EXACT_TOKENS_IN_FOR_BPT_OUT:
+         * User sends precise quantities of tokens, and receives an
+         * estimated but unknown (computed at run time) quantity of BPT.
+         *
+         * ['uint256', 'uint256[]', 'uint256']
+         * [EXACT_TOKENS_IN_FOR_BPT_OUT, amountsIn, minimumBPT]
+         */
+        bytes memory userData = abi.encode(
+            IBalancerVault.WeightedPoolJoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT,
+            amountsIn,
+            minBPTwSlippage
+        );
+
+        IBalancerVault.JoinPoolRequest memory request = IBalancerVault
+            .JoinPoolRequest(poolAssets, amountsIn, userData, false);
+
+        // Add the pool assets in this strategy to the balancer pool
+        balancerVault.joinPool(
+            balancerPoolId,
+            address(this),
+            address(this),
+            request
+        );
+
+        // Deposit the Balancer Pool Tokens (BPT) into Aura
+        _lpDepositAll();
+    }
+
+    /**
+     * @notice Withdraw a Vault collateral asset from the Balancer pool.
+     * @param _recipient Address to receive the Vault collateral assets. Typically is the Vault.
+     * @param _asset Address of the Vault collateral asset
+     * @param _amount The amount of Vault collateral assets to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override whenNotInVaultContext onlyVault nonReentrant {
+        address[] memory assets = new address[](1);
+        uint256[] memory amounts = new uint256[](1);
+        assets[0] = _asset;
+        amounts[0] = _amount;
+
+        _withdraw(_recipient, assets, amounts);
+    }
+
+    /**
+     * @notice Withdraw multiple Vault collateral asset from the Balancer pool.
+     * @param _recipient Address to receive the Vault collateral assets. Typically is the Vault.
+     * @param _assets Addresses of the Vault collateral assets
+     * @param _amounts The amounts of Vault collateral assets to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address[] memory _assets,
+        uint256[] memory _amounts
+    ) external whenNotInVaultContext onlyVault nonReentrant {
+        _withdraw(_recipient, _assets, _amounts);
+    }
+
+    /**
+     * @dev Withdraw multiple Vault collateral asset from the Balancer pool.
+     * @param _recipient Address to receive the Vault collateral assets. Typically is the Vault.
+     * @param _assets Addresses of the Vault collateral assets
+     * @param _amounts The amounts of Vault collateral assets to withdraw
+     */
+    function _withdraw(
```

Comment:
This technique currently used by this method is fairly ugly. On a ten million dollar withdraw, this could leave $10,000 in limbo. And that's if the oracles are working perfectly and meshed with the pool prices. We might have to use looser oracle numbers.

What we really want to know is the exact number of BPT tokens we need to withdraw before we get them from staking.

When a withdraw happens, the balancer vault asks the balancer pool for the details of what should happen, which ends up including the number of BPT tokens to burn. This is calculated internally to the pool based on the requested tokens.

The math actually isn't that bad:

<img width="991" alt="image" src="https://github.com/OriginProtocol/origin-dollar/assets/837/7472beb5-5e29-41c3-8b33-1aa9c88f924d">

(Though we would need to include the fee adjustments before running this method, which doesn't look nearly as hard as curve.)

While it would be nice if they had an external interface on this, I don't see it in the pool contract. We could just make a little helper contract that runs their version of solidity and has the needed external interface to run this method.

Reply:
Or, if we are going to withdraw more tokens than needed, we could use _lpDepositAll() at the end to deposit them back to earning.

We could also a simple max cap on the maxBPTtoWithdraw to our current balance - if we estimate more than we have, we just use the amount that we have. This would allow us to withdraw right up to the edge of what we can.

Reply:
We could also skip all the estimation, and withdraw all but one raw LP token (gas savings), do the withdraw, then put it all back.

Reply:
Exact amounts are probably too much work, on reflection.

Reply:
I've addressed this issue to some extent in another comment. So the problem is that with:
1. exact oracle prices -> as you mention we are still leaving 0.1% on the table
2. oracles lagging, assets more expensive -> we wouldn't calculate sufficient BPT and it will revert
3. oracles lagging, assets cheaper -> we are exposed to being MEVed

Since all the prices are ETH denominated I wouldn't expect big moves to making assets more expensive. On the other hand there could be slashing (point 3.) events that would make assets cheaper and be within oracle tolerances and make us exposed to MEV.

One reservation regarding deriving BPT strictly off of the pool I have is that that number is manipulatable with flash loans tilting the pool.

I need to do some research how rETH/ETH wstETH/ETH oracle prices behave. I would expect them to very slowly move up. In that case our strategy will mostly perform the scenario 1. and leave some extra BPT in the strategy. 

The real danger is the slashing of underlying LSDs. One (temporary) way to mitigate this would be to have it built into our off-chain allocation script, that if Oracle prices and pool balances (price) deviate too much, the allocation doesn't get built (to not lose funds with MEV). Need to think more on this one. 


And you are right the BPT math is pretty straightforward. Tried to cram it into the strategy contract, but it reaches the deployment size limit -> https://github.com/OriginProtocol/origin-dollar/pull/1740 Like you said it would need to be deployed as a separate contract.